### ISSUE TITLE
Support statically shaped dynamic operations (nonzero, unique, etc)

### ISSUE THREAD

**Issue by zou3519** [pytorch/functorch#256]
## 🚀 Feature

Cross-post of https://github.com/pytorch/pytorch/issues/62320.

torch.nonzero, torch.unique, etc should accept an optional size argument to produce statically shaped output

## Motivation

(1) Make it so that they can be vmapped over
(2) Make it easier to lower them to backends

## Pitch

Either:
- add a variant of these operations that accepts a `size=None` argument
- add `size=None` to these operations.

## Alternatives

N/A

## Additional context

N/A

**Comment by zou3519** [comment]
This is actionable: we should (1) identify all of the dynamic shaped operations in PyTorch (2) propose a modification to them (3) get buy-in and (4) implement and ship

**Comment by sergiynesterenko90** [comment]
Also hitting the error that points to this ticket when applying vmap to a function that uses torch.where

**Comment by enijkamp** [comment]
+1, trying to vmap() over unique()

**Comment by ezyang** [comment]
We were discussing this issue in issue triage, and one thing that wasn't clear to us was whether or not a size=... argument would actually help (as it would change what you would get, e.g., you would have to handle the resulting fill value if you didn't have enough entries, or you would have to be OK with truncating elements in that case.) So it's not a drop-in replacement, you have to know a bit about what exactly you are using nonzero/unique/where for.

Can folks here describe in more detail what they are using these operations for / post their model code? That would help us a lot in prioritizing this!

**Comment by ezyang** [comment]
Yeah, so for your first case, size doesn't help. If the computations on the branches were cheap, you could just do R * 2 and R * 0.5 and then use torch.where to select between the results in one computation or the other. Even if they're not too cheap, with torch.compile we could fuse these into a single kernel to avoid memory bandwidth.

But if the compute on R is sufficiently expensive that it is profitable to avoid running it on the entire batch, then I think you have no choice to do a DtoH sync. But that's fine maybe: we aren't here to avoid DtoH sync, we just wanted to make vmap work. I think we can make it work: I think we could augment torch.cond to do what you want! If torch.cond has a data dependent condition, the way this should desugar is into a DtoH syncing operation to partition into the left and right sides, and then running batched operations on the branches, and then cat'ing them back together (we probably should preserve ordering, which means you need a more expensive interleaving operation). But in any case size doesn't help.

I agree that size for unique seems perfect in the example you gave.

cc @zou3519 @Chillee

**Comment by ezyang** [comment]
Well, vmap should be expressible as a source to source transform, right? So we just need to make sure that it transforms to something that we can keep going with. If that's nested tensor, then fine. But maybe there's something easier; it doesn't feel like this example should be worse with two batch dims rather than one.

**Comment by AlphaBetaGamma96** [comment]
Hi All, I just had a look at my example above and I've resolved the issue. The problem was that I didn't specify the `argnums` in the `jacrev` call, so I assume it tried to `jacrev` over both the derivatives with respect to my input and the parameters. If I specify the `argnums`, it works fine. For reference, I'm using torch version `2.0.0`.

**Comment by kanbei7** [comment]
Just a thought, there might be a temporary workaround for unique, that is to implement your own "unique" function. The down side is that it involves sort. Assume a 2D tensor X, where X[0,:] = 0.

```
def handmade_unique(x):
    sorted_x, _ = torch.sort(x)
    mask = (sorted_x[1:] - sorted_x[:-1])>0
    return sorted_x[1:]*mask
```

This works well with vmap for me.
```
padded_X = F.pad(X,(1,0))
new_X = torch.vmap(handmade_unique)(padded_X)
```

**Comment by aboubezari** [comment]
Would love to see this support for torch NMS: https://pytorch.org/vision/main/generated/torchvision.ops.nms.html#torchvision.ops.nms

**Comment by ezyang** [comment]
@aboubezari put this request in the vision repo!
